Fix bug 1530988 - Download project config files instead of using the …

jotes opened the pull request:
TODO:
* [x] - Check the logic
* [x] - Fix failing CI
* [x] - Write tests

jotes commented on pontoon/sync/vcs/models.py:
```diff
@@ -37,6 +40,37 @@
+    def parse(self, path, env=None, ignore_missing_includes=True):
+        config_path = path.replace(self.checkout_path, "")
+
+        if "{locale_code}" in self.permalink_prefix:
```
@mathjazz Howdy!

Can you check if the logic behind `remote_path` and `download_path` is okay? I tested the code against the `www-l10n` repo and it seems to work okay.

mathjazz replied:
I think it is! Left some more notes.

mathjazz commented on pontoon/sync/vcs/models.py:
```diff
@@ -37,6 +40,37 @@
+        self.permalink_prefix = permalink_prefix
+
+    def parse(self, path, env=None, ignore_missing_includes=True):
+        config_path = path.replace(self.checkout_path, "")
```
Do we need that `replace()`?

IIUC, the `path` is `self.configuration_file`, which doesn't include the `self.checkout_path`.

jotes replied:
Unfortunately, that's not so simple :/ When TOMLParser parses the main project file and it contains in the project config files, `parse()` is called again for every referenced file and passes its absolute path.

mathjazz commented on pontoon/sync/vcs/models.py:
```diff
@@ -37,6 +40,37 @@
+        config_path = path.replace(self.checkout_path, "")
+
+        if "{locale_code}" in self.permalink_prefix:
+            remote_path = (self.permalink_prefix or "").format(locale_code=config_path)
```
If the above condition is `True`, we don't need `or ""`, right?

mathjazz commented on pontoon/sync/vcs/models.py:
```diff
@@ -37,6 +40,37 @@
+        if "{locale_code}" in self.permalink_prefix:
+            remote_path = (self.permalink_prefix or "").format(locale_code=config_path)
+        else:
+            remote_path = urljoin(self.permalink_prefix or "", config_path)
```
I can't see how this would work if `self.permalink_prefix` is None. If it is, I think we should raise an error.

mathjazz commented on pontoon/sync/vcs/models.py:
```diff
@@ -114,6 +158,10 @@ def __init__(
         self.configuration = None
         if db_project.configuration_file:
+            # Permalink is required to download project config files.
+            if db_project.source_repository.permalink_prefix:
```
`if not`?

jotes replied:
`if not` is on my local machine already :D Sorry for pushing partial changes. I'll ping you when it's ready for the next stage of review.

mathjazz commented on Makefile:
```diff
@@ -75,6 +76,9 @@ lint-frontend:
 
+sync-projects:
+	"${DC}" run --rm webapp .//manage.py sync_projects $(opts)
+
```
Let's keep the order of items here the same as below.

And let's also document the missing targets: `flake8` and `test-frontend`.

mathjazz commented on pontoon/sync/vcs/models.py:
```diff
@@ -37,6 +40,54 @@
+class MissingRepositoryPermalink(Exception):
+    """
+    Raised when a project uses project config files and
+    and its source repository doesn't have the permalink.
```
Nit: Double and.

mathjazz commented on pontoon/sync/vcs/models.py:
```diff
@@ -527,9 +582,10 @@ def l10n_base(self):
     def parsed_configuration(self):
         """Return parsed project configuration file."""
-        return TOMLParser().parse(
-            self.configuration_path, env={"l10n_base": self.l10n_base},
```
That means we can also remove other traces of `configuration_path` from this file and the test file.

mathjazz commented on pontoon/sync/vcs/models.py:
```diff
@@ -37,6 +40,54 @@
+
+    def get_project_config(self, path):
+        """Download the project config file and return its local path."""
+        download_path = self.get_download_path(path)
```
Nit: I'd suggest we call it `local_path` since the comment calls it that way and download path could mean remote path as well.

mathjazz commented on pontoon/sync/vcs/models.py:
```diff
@@ -37,6 +40,54 @@
+        """Download the project config file and return its local path."""
+        download_path = self.get_download_path(path)
+
+        with open(download_path, "wb") as f:
```
We need to make sure we create the file if it doesn't exist yet. Most of the time it will NOT exist on the `web` worker on prod and will raise an error similar to this one:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.8/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/app/pontoon/base/views.py", line 642, in download
    content, filename = utils.get_download_content(slug, code, part)
  File "/app/pontoon/base/utils.py", line 238, in get_download_content
    source_path = _download_file(
  File "/app/pontoon/base/utils.py", line 144, in _download_file
    absolute_l10n_path = vcs_project.configuration.l10n_path(
  File "/app/pontoon/sync/vcs/models.py", line 653, in l10n_path
    project_files = self.get_or_set_project_files(locale.code)
  File "/app/pontoon/sync/vcs/models.py", line 641, in get_or_set_project_files
    and locale_code not in self.parsed_configuration.all_locales
  File "/usr/local/lib/python3.8/site-packages/django/utils/functional.py", line 80, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/app/pontoon/sync/vcs/models.py", line 590, in parsed_configuration
    return DownloadTOMLParser(
  File "/app/pontoon/sync/vcs/models.py", line 85, in parse
    self.get_project_config(path), env, ignore_missing_includes
  File "/app/pontoon/sync/vcs/models.py", line 74, in get_project_config
    with open(download_path, "wb") as f:
FileNotFoundError: [Errno 2] No such file or directory: '/app/media/projects/project-config/mathjazz/moz-l10n-config-test.git/l10n.toml'
```

mathjazz replied:
Ideally, we should only download the config file if it's not available yet.

mathjazz reviewed (commented):
Thanks for the update!

Left some notes.

mathjazz commented:
I have a project set up against https://github.com/mathjazz/android-l10n.

I set https://raw.githubusercontent.com/mathjazz/android-l10n/master/l10n.toml as the download prefix.

When I try to download the file, I get the following error:
```
400 Client Error: Bad Request for url: https://raw.githubusercontent.com/mozilla-mobile/fenix/l10n.toml
```

mathjazz commented on pontoon/sync/vcs/models.py:
```diff
@@ -37,7 +41,63 @@
+
+
 class MissingSourceRepository(Exception):
+
```
Nit: you can remove the blank line.

jotes commented:
Hey @mathjazz,
I removed many chunks of my changes and fixed the bug in the previous version of the code.

Could you look at this and test it on Staging? I'm able to download the resources locally from the projects you gave me.

mathjazz commented on pontoon/base/utils.py:
```diff
@@ -214,13 +215,19 @@ def get_download_content(slug, code, part):
+            project.repositories.filter(Q(permalink_prefix__contains="{locale_code}"))
             .values_list("permalink_prefix", flat=True)
             .distinct()
         )
```
Can we only make the distinction (based on presence of `project.configuration_file`) for the QS, and then apply `values_list()` and `distinct()` in one place only?

Also, can we omit that `Q()` wrapper?

mathjazz reviewed (commented):
Thanks for the update!

I've deployed to stage and tested the following projects:

- Android L10n
- Mozilla.org
- Privacy Not Included (buyersguide.toml)
- Mozilla Festival (mozfest.toml)
- Foundation website (foundation.toml)
- Donate - Mozilla content
- Donate - Thunderbird content

Everything works fine!!

Except for the last project (Donate - Thunderbird content), which raises an error:

```

Message: ValueError: no path specified
at line 113, /app/.heroku/python/lib/python3.8/site-packages/django/core/handlers/base.py,
 
113. _get_response response = wrapped_callback(request, *callback_args, **callback_kwargs)
at line 544, /app/.heroku/python/lib/python3.8/site-packages/newrelic/hooks/framework_django.py,
 
544. wrapper return wrapped(*args, **kwargs)
at line 75, /app/.heroku/python/lib/python3.8/contextlib.py,
 
75. inner return func(*args, **kwds)
at line 669, /app/pontoon/base/views.py,
 
669. download content, filename = utils.get_download_content(slug, code, part)
at line 232, /app/pontoon/base/utils.py,
 
232. get_download_content locale_path = _download_file(
at line 148, /app/pontoon/base/utils.py,
 
148. _download_file relative_l10n_path = os.path.relpath(
at line 453, /app/.heroku/python/lib/python3.8/posixpath.py,
 
453. relpath raise ValueError("no path specified") 
```

That's how the repo config for that project (enabled for `fr`) looks like:
<img width="1042" alt="Screenshot 2020-09-20 at 22 56 05" src="https://user-images.githubusercontent.com/626716/93722084-8bf45a80-fb94-11ea-80f9-4b5187bc2e80.png">

mathjazz commented:
I've also noticed that localizable Gettext files are downloaded with file extension POT instead of PO.

mathjazz commented on pontoon/sync/tests/test_vcs_models.py:
```diff
@@ -297,11 +334,10 @@ def setUp(self):
-        self.vcs_project.configuration.configuration_path = os.path.join(
-            PROJECT_CONFIG_CHECKOUT_PATH, self.db_project.configuration_file,
+        self.db_project.source_repository.permalink_prefix = (
+            "https://example.com/{locale_code}"
```
I'd remove `{locale_code}` to avoid confusion: we don't use {locale_code} in projects that use configuration files.

mathjazz commented on pontoon/sync/tests/test_vcs_models.py:
```diff
@@ -417,17 +453,18 @@ def setUpResource(self):
-    self.vcs_project.configuration.configuration_path = os.path.join(
-        PROJECT_CONFIG_CHECKOUT_PATH, self.db_project.configuration_file,
+    self.db_project.source_repository.permalink_prefix = (
+        "https://example.com/{locale_code}"
```
Same here.

mathjazz reviewed (commented):
Thanks for the update!

Some nits and we're done!

mathjazz reviewed (approved):
\o/

Thanks for all the work here! This is an important step in making project config usable in more critical projects.

I've rebased against master to resolve the conflict in `Makefile`.
